src/MonoTorrent/MonoTorrent.Client/Managers/TrackerManager.cs

```diff
@@ -108,80 +108,114 @@ internal TrackerManager (ITrackerRequestFactory requestFactory, IEnumerable<RawT
         #region Methods
 
         public async Task Announce ()
-        {
-            await Announce (TorrentEvent.None);
-        }
+            => await Announce (TorrentEvent.None);
 
         public async Task Announce (TorrentEvent clientEvent)
         {
-            await Announce (clientEvent, null);
+            // If the user initiates an Announce we need to go to the correct thread to process it.
+            await ClientEngine.MainLoop;
+
+            var announces = new List<Task> ();
+            for (int i = 0; i < Tiers.Count; i++) {
+                var tier = Tiers[i];
+                var tracker = tier.ActiveTracker;
+                var interval = tier.LastAnnounceSucceeded ? tracker.UpdateInterval : tracker.MinUpdateInterval;
+                if (tier.LastAnnounce.Elapsed > interval && (clientEvent != TorrentEvent.Stopped || tier.LastAnnounceSucceeded))
+                    announces.Add (AnnounceToTier (clientEvent, tier));
+            }
+            await Task.WhenAll (announces);
         }
 
         public async Task Announce (ITracker tracker)
         {
             Check.Tracker (tracker);
-            await Announce (TorrentEvent.None, tracker);
-        }
 
-        async Task Announce (TorrentEvent clientEvent, ITracker referenceTracker)
-        {
             // If the user initiates an Announce we need to go to the correct thread to process it.
             await ClientEngine.MainLoop;
+            await Announce (TorrentEvent.None, tracker);
+        }
 
-            LastAnnounce.Restart ();
-            LastUpdated = DateTime.UtcNow;
-
-            AnnounceParameters p = RequestFactory.CreateAnnounce (clientEvent);
-
-            foreach ((TrackerTier trackerTier, ITracker tracker) in GetNextTracker (referenceTracker)) {
-                try {
-                    // If we have not announced to this Tracker tier yet then we should replace the ClientEvent.
-                    // But if we end up announcing to a different Tracker tier we may want to send the
-                    // original/unmodified args.
-                    AnnounceParameters actualArgs = p;
-                    if (!trackerTier.SentStartedEvent)
-                        actualArgs = actualArgs.WithClientEvent (TorrentEvent.Started);
-
-                    List<Peer> peers = await tracker.AnnounceAsync (actualArgs);
-                    LastAnnounceSucceeded = true;
-                    AnnounceComplete?.InvokeAsync (this, new AnnounceResponseEventArgs (tracker, true, peers.AsReadOnly ()));
+        async Task AnnounceToTier (TorrentEvent clientEvent, TrackerTier tier)
+        {
+            for (int i = 0; i < tier.Trackers.Count; i++) {
+                int trackerIndex = (i + tier.ActiveTrackerIndex) % tier.Trackers.Count;
```

Comment:
Could you please add a comment, I'm not sure I'm following logic here on how index is calculated

Reply:
Suppose there are 10 trackers in the `TrackerTier` and `tracker[4]` was the one which responded most recently. This logic ensures we try each tracker sequentially `trackerIndex` will be calculated as:
```
(0 + 4) % 10   // 4
(1 + 4) % 10   // 5
(2 + 4) % 10   // 6
(3 + 4) % 10   // 7
(4 + 4) % 10   // 8
(5 + 4) % 10   // 9
(6 + 4) % 10   // 0
(7 + 4) % 10   // 1
(8 + 4) % 10   // 2
(9 + 4) % 10   // 3
```